### ISSUE TITLE
Hard to use with CSS-in-JS libs.

### ISSUE THREAD

**Issue by whmountains** [netlify/netlify-cms#793]
## Background

I'm currently using Netlify CMS to build a website for a client.  It's a completely static site, but it uses React as a template engine.  I also use the React bundle to render a preview inside Netlify CMS.

The other important detail is that this website is styled with [styled-components](https://www.styled-components.com/), which works by injecting css rules into `document.getElementsByTagName('head')[0]`.

## The Problem

Netlify CMS [transplants](https://github.com/netlify/netlify-cms/blob/54476cccd32d4edb8f6805b1bf0640463fec1e0d/src/components/ScrollSync/ScrollSyncPane.js#L18) the preview component inside an iframe, while the `<style>` elements generated by `styled-components` remain outside the iframe.  This leaves me with an unstyled preview, which is most unappealing. ✨

I haven't tested, but I expect this to affect other CSS-in-JS libraries like [Glamorous](https://glamorous.rocks/), [jsxstyle](https://github.com/smyte/jsxstyle), or [JSS](https://github.com/cssinjs/jss) as well as any other React library that injects extra elements, like [react-helmet](https://github.com/nfl/react-helmet) or [react-portal](https://github.com/tajo/react-portal).

**Comment by erquhart** [comment]
@whmountains it's not always a given that a site's CSS will automatically apply to the preview pane - doing so often requires porting the styles in via [`registerPreviewStyle`](https://www.netlifycms.org/docs/customization/#registerpreviewstyle). Does styled-components provide any way to output a CSS file?

**Comment by erquhart** [comment]
Hmm looks like styled components adds a data attribute to its style elements - I'll bet the others do the same. The `registerPreviewStyle` registry accepts file paths only, but it could also accept a CSS selector for this use case, which we could run with `querySelectorAll` and copy matching elements into the preview pane. We should also accept raw styles while we're on the subject.

That said, we need to give some higher level consideration to the proper abstraction for these API changes. What do you think?

**Comment by zionis137** [comment]
I had the same problem with CSS-Modules on GatsbyJS, I hope this helps:

according to the [documentation](https://www.npmjs.com/package/style-loader#insertinto) style-loader is able to inject the inline-CSS into an iframe. 

But in the end I was unable to set up this functionality with netlify-cms and used the [Extract-Text-Plugin](https://www.npmjs.com/package/extract-text-webpack-plugin)
This extracts every CSS from all components into a new stylesheet which i include with
`CMS.registerPreviewStyle('stylesCMS.css')`


The relevant parts from my `webpack.config` look like this:
```
const ExtractTextPlugin = require("extract-text-webpack-plugin");
const extractLess = new ExtractTextPlugin({
    filename: "stylesCMS.css",
    // path: path.resolve(__dirname, '../static/admin/')
});

.... 
  module: {
    rules: [
  {
         test: /\.css$/,
         use: extractLess.extract({
           use: [
           {
               loader: "css-loader?modules&importLoaders=1"
           }],
         })
```

**Comment by whmountains** [comment]
Nice trick @zionis137.

I'm still hoping for some more official support, but this is a nice workaround!

**Comment by whmountains** [comment]
@erquhart Your proposal for finding CSS by selector sounds reasonable.

Another possibility would be to allow loading the preview iframe via a URL rather than trying to inject a react tree inside.  It would be a fairly trivial change and IMO completely solve the problem.

**Comment by tizzle** [comment]
Hey,

i'm wondering if there is any news on this, as i ran into the exact same issue as @whmountains. In addition i feel that using `extract-text-webpack-plugin` is not going to work, as in my understanding this won't pick up the styled-components definitions. This is discussed [here](https://github.com/styled-components/styled-components/issues/1018), [here](https://stackoverflow.com/questions/44916599/how-can-i-opt-for-react-styled-components-to-generate-a-physical-css-file) and [here](https://spectrum.chat/thread/6b445df6-c807-472d-820b-f161bcacbd8e) in a little more detail.

Maybe rendering into a React portal instead of an iFrame would solve the issue?

**Comment by erquhart** [comment]
That's an interesting idea. I haven't looked into portals at all, but feel free to check it out and see if it's possible.

**Comment by whmountains** [comment]
@erquhart To answer your question I would create a subscriber interface, which a HoC within the preview pane can access.  Similar to how react-redux works with `connect` wrapping the `store.subscribe` api.  In fact, I would copy the react-redux api as much as possible since it's performant and everyone knows how to use it.

You could also have another HoC which would wrap the entire preview pane and implement scroll sync by listening to scroll events.

**Comment by whmountains** [comment]
AFAIK netlify-cms uses redux under the hood.  Could you just expose the store inside the iframe?

Everything from EditorPreviewPane on down would be incorporated into the custom build running inside the iframe.

Just throwing out ideas.  I'm not very familiar with the codebase or all the caveats a system like this would introduce.  It just seems that netlify-cms's preview interface is breaking core assumptions about how web pages are rendered and it would be nice to fix that.

**Comment by robertgonzales** [comment]
For anyone using [emotion](https://emotion.sh/), I solved this issue by using SSR on the fly to extract the css and then inject it into the nearest document (iframe) head. Very hacky but it works.

```js
import { renderToString } from "react-dom/server"
import { renderStylesToString } from "emotion-server"

class CSSInjector extends React.Component {
  render() {
    return (
      <div
        ref={ref => {
          if (ref && !this.css) {
            this.css = renderStylesToString(renderToString(this.props.children))
            ref.ownerDocument.head.innerHTML += this.css
          }
        }}>
        {React.Children.only(this.props.children)}
      </div>
    )
  }
}
```
It works by wrapping your preview template:
```js
CMS.registerPreviewTemplate("blog", props => (
  <CSSInjector>
    <BlogPreviewTemplate {...props />
  </CSSInjector>
))

**Comment by erquhart** [comment]
That will be much less hacky once #1162 lands. Any library that can export strings will be covered at that point.

Anyone up for talking @maxstbr into exporting strings from styled components?

**Comment by erquhart** [comment]
On second thought that PR won't help for the emotion case at all. I'm also thinking that what you've done really isn't hacky, especially considering how you moved it into a component. This might find it's way into the docs lol

**Comment by mxstbr** [comment]
styled-components has a way to target an iframe <head> as it's injection point:

```JSX

import { StyleSheetManager } from 'styled-components'

<StyleSheetManager target={iframeHeadElem}>
  <App />
</StyleSheetManager>
```

Any styled component within `App` will now inject it's style tags into the target elem! Maybe that's helpful?

**Comment by erquhart** [comment]
I was just looking at `StyleSheetManager` recently and wondering if it might work for this - looks like it should!

@whmountains care to give it a go and let us know?

**Comment by mxstbr** [comment]
Note that the target feature was only introduced in v3.2.0 and isn't documented just yet: https://www.styled-components.com/releases#v3.2.0_stylesheetmanager-target-prop

**Comment by markacola** [comment]
I just tried this out and it works great! Just simply:
```javascript
  const iframe = document.querySelector(".nc-previewPane-frame")
  const iframeHeadElem = iframe.contentDocument.head;

  return (
    <StyleSheetManager target={iframeHeadElem}>
      {/* styled elements */}
    </StyleSheetManager>
  )
```

**Comment by erquhart** [comment]
Leaving this open as I'd still like to discuss how our API might improve so that this isn't so manual. Perhaps we need some kind of preview plugin API that would allow a styled-components plugin to handle this behind the scenes.

**Comment by Firthir** [comment]
We got this working. It will bring in all styles from the front end URL.
```
if (
  window.location.hostname === 'localhost' &&
  window.localStorage.getItem('netlifySiteURL')
) {
  CMS.registerPreviewStyle(
    window.localStorage.getItem('netlifySiteURL') + '/styles.css'
  )
} else {
  CMS.registerPreviewStyle('/styles.css')
}

CMS.registerPreviewTemplate('home-page', ({ entry }) => (
  <HomePageTemplate {...entry.toJS().data} />
))
```

**Comment by pungggi** [comment]
@Firthir Can you elaborate what is the idea behind the localStorage Item?

**Comment by Firthir** [comment]
Hey @pungggi you'll find this useful if you're using localhost and logging into the CMS. This get the preview to display correctly locally. I use the .env.development var for a few things.
Try this, then we can use this with the example above.
- Create a new .env.development file in your root dir.
- add the ```NETLIFY_SITE_URL=https://yoursstiehere.netlify.com```

OR in this example just hardcode this into your custom JS in the admin/index.html file.
```
if (
  window.location.hostname === 'localhost' 
) {
  CMS.registerPreviewStyle( 'https://yoursstiehere.netlify.com/styles.css')
} else {
  CMS.registerPreviewStyle('/styles.css')
}
```
Hope that helps you or someone else.

**Comment by erquhart** [comment]
@Firthir this issue is about CSS in JS solutions like Emotion, Styled Components, etc.

**Comment by jonathanphz** [comment]
I tried to use @robertgonzales solution but get error `Objects are not valid as a React child ` :(

I'm using:
    "gatsby": "next",
    "gatsby-plugin-emotion": "^2.0.0-rc.0",
    "gatsby-plugin-netlify": "^2.0.0-rc.0",
    "gatsby-plugin-netlify-cms": "^3.0.0-beta.0",

**Comment by pungggi** [comment]
https://github.com/netlify/netlify-cms/issues/1408#issuecomment-424965185



> 
> 
> I just tried this out and it works great! Just simply:
> 
> ```js
>   const iframe = document.querySelector(".nc-previewPane-frame")
>   const iframeHeadElem = iframe.contentDocument.head;
> 
>   return (
>     <StyleSheetManager target={iframeHeadElem}>
>       {/* styled elements */}
>     </StyleSheetManager>
>   )
> ```

Seems not to work with v2 
https://github.com/netlify/netlify-cms/issues/1408#issuecomment-424965185

**Comment by joserocha3** [comment]
Replace the first line with
```javascript
const iframe = document.getElementsByTagName('iframe')[0]
```

**Comment by MiniCodeMonkey** [comment]
> For anyone using [emotion](https://emotion.sh/), I solved this issue by using SSR on the fly to extract the css and then inject it into the nearest document (iframe) head. Very hacky but it works.
>

This worked great! Just make sure to import `react` as well.

> ```js
> import React from "react"
> import { renderToString } from "react-dom/server"
> import { renderStylesToString } from "emotion-server"
> 
> class CSSInjector extends React.Component {
>   render() {
>     return (
>       <div
>         ref={ref => {
>           if (ref && !this.css) {
>             this.css = renderStylesToString(renderToString(this.props.children))
>             ref.ownerDocument.head.innerHTML += this.css
>           }
>         }}>
>         {React.Children.only(this.props.children)}
>       </div>
>     )
>   }
> }
> ```
> 
> It works by wrapping your preview template:
> 
> ```js
> CMS.registerPreviewTemplate("blog", props => (
>   <CSSInjector>
>     <BlogPreviewTemplate {...props} />
>   </CSSInjector>
> ))
> ```

**Comment by unyo** [comment]
Trying to figure out how pull in CSS styles via netlify-cms using @emotion/styled v10, came up with this solution. Apparently netlify-cms just overwrites the whole head with "cms.css", and the renderToString + ref method doesn't fully generate the styles for some reason.

(This is a terrible hack imo - I'd like to update netlify-cms to preserve the existing head contents instead somehow)

```javascript
import React from "react"

export default class CSSInjector extends React.Component {
  componentDidMount() {
    if (!this.css) {
      const el = document.getElementsByTagName('head')[0]
      const iframeDoc = document.getElementById('nc-root').getElementsByTagName('iframe')[0].contentDocument
      const parentEl = iframeDoc.body.parentElement
      const newHead = document.createElement('head')
      this.css = el.innerHTML
      newHead.innerHTML = this.css
      parentEl.appendChild(newHead)
    }
  }
  render() {
    return (
      <div id="css-injector">
        {React.Children.only(this.props.children)}
      </div>
    )
  }
}
```

**Comment by asistapl** [comment]
@robertgonzales  Hey, now with emotion 10 we have to do something like this:

```jsx
class CSSInjector extends React.Component {
  constructor() {
    super()
    const iframe = document.getElementsByTagName('iframe')[0]
    const iframeHead = iframe.contentDocument.head
    this.cache = createCache({ container: iframeHead })
  }

  render() {
    return (
      <CacheProvider value={this.cache}>
        {this.props.children}
      </CacheProvider>
    )
  }
}
```

**Comment by Tomekmularczyk** [comment]
`document.getElementsByTagName('iframe')[0]` didn't work for me. In my case for styled-components I had to do:

```js
import React from "react";
import { StyleSheetManager } from "styled-components";

export default function StyledSheets({ children }) {
  const iframe = document.querySelector("#nc-root iframe");
  const iframeHeadElem = iframe && iframe.contentDocument.head;

  if (!iframeHeadElem) {
    return null;
  }

  return (
    <StyleSheetManager target={iframeHeadElem}>{children}</StyleSheetManager>
  );
}
```

**Comment by homearanya** [comment]
Hi there, 

Has anyone been successful with Material-UI? Following similar approaches described above I got to this:
```javascript
import React from 'react';
import { install } from '@material-ui/styles';
import { StylesProvider } from '@material-ui/styles';
import { jssPreset } from '@material-ui/core/styles';
import { create } from 'jss';
import camelCase from 'jss-plugin-camel-case';

install();

export default function MaterialUiSheets({ children }) {
  const iframe = document.querySelector('#nc-root iframe');
  const iframeHeadElem = iframe && iframe.contentDocument.head;

  if (!iframeHeadElem) {
    return null;
  }

  const jss = create({
    plugins: [...jssPreset().plugins, camelCase()],
    insertionPoint: iframeHeadElem.firstChild,
  });

  return <StylesProvider jss={jss}>{children}</StylesProvider>;
}

```
This only works partially, meaning not all the styles are injected into the Iframe.

This is probably a Material-UI question, which I'm not familiar with, but anyway someone might be able to help.

**Comment by devolasvegas** [comment]
Hi All! 

I struggled with this a bit. For some reason setting this up (with styled-components ^4.2.0, netlify-cms-app 2.9.1, and netlify-cms-core 2.11.0) as a functional component wouldn't work, so I tried using the effect hook, which didn't work either. But perhaps my grasp of the effect hook isn't what it should be. Ended up setting it up as a class component which worked mostly, except for my theme.

So if you are using a theme with styled-components, you should be able to get this to work simply by wrapping your StyleSheetManager component with the ThemeProvider component. Perhaps there is a better way to do this, but I have spent enough time on this for now, and am moving on!

```import React, { Component } from 'react';
import { StyleSheetManager, ThemeProvider } from 'styled-components';

import theme from '../styles/theme/theme';

class StylesheetInjector extends Component {
	constructor(props) {
		super(props);
		this.state = {
			iframeRef: '',
		};
	}

	componentDidMount() {
		const iframe = document.querySelector('#nc-root iframe');
		const iframeHeadElem = iframe && iframe.contentDocument.head;
		this.setState({ iframeRef: iframeHeadElem });
	}

	render() {
		return (
			<>
				{this.state.iframeRef && (
					<ThemeProvider theme={theme}>
						<StyleSheetManager target={this.state.iframeRef}>
							{this.props.children}
						</StyleSheetManager>
					</ThemeProvider>
				)}
			</>
		);
	}
}

export default StylesheetInjector;`

**Comment by devolasvegas** [comment]
Refactored using hooks:

```javascript
import React, { useState, useEffect } from 'react';
import { StyleSheetManager, ThemeProvider } from 'styled-components';

import theme from '../styles/theme/theme';

const StylesheetInjector = ({ children }) => {
	const [iframeRef, setIframeRef] = useState(undefined);

	useEffect(() => {
		const iframe = document.querySelector('#nc-root iframe');
		const iframeHeadElem = iframe && iframe.contentDocument.head;
		setIframeRef(iframeHeadElem);
	});

	return (
		<>
			{iframeRef && (
				<ThemeProvider theme={theme}>
					<StyleSheetManager target={iframeRef}>{children}</StyleSheetManager>
				</ThemeProvider>
			)}
		</>
	);
};

export default StylesheetInjector;
```

**Comment by erquhart** [comment]
Did that do the trick?

**Comment by devolasvegas** [comment]
Indeed, it did. I am also using createGlobalStyle and found that my global style component would work with this if I placed it inside the StyleSheetManager component, adjacent to { children }.

**Comment by d3sandoval** [comment]
@homearanya I was able to get material-ui styles to appear in the iFrame by following this approach: https://github.com/mui-org/material-ui/issues/13625#issuecomment-493608458

I ended up with the following, as an example:
```
import React from 'react'
import PropTypes from 'prop-types'
import CssBaseline from '@material-ui/core/CssBaseline';
import { ThemeProvider } from '@material-ui/styles';
import { create } from "jss";
import { jssPreset, StylesProvider } from "@material-ui/styles";
import theme from '../../theme';
import { AboutPageTemplate } from '../../templates/about-page'

class AboutPagePreview extends React.Component {
  state = {
    ready: false
  };

  handleRef = ref => {
    const ownerDocument = ref ? ref.ownerDocument : null;
    this.setState({
      ready: true,
      jss: create({
        ...jssPreset(),
        insertionPoint: ownerDocument.querySelector("#demo-frame-jss")
      }),
      sheetsManager: new Map()
    });
  };

  render() {
    const { entry, widgetFor } = this.props;
    const data = entry.getIn(['data']).toJS()

    if (data) {
      return (
        <React.Fragment>
          <div id="demo-frame-jss" ref={this.handleRef} />
          {this.state.ready ? (
          <StylesProvider
            jss={this.state.jss}
            sheetsManager={this.state.sheetsManager}
          >
            <ThemeProvider theme={theme}>
              {/* CssBaseline kickstart an elegant, consistent, and simple baseline to build upon. */}
              <CssBaseline />
              <AboutPageTemplate
                title={entry.getIn(['data', 'title'])}
                content={widgetFor('body')}
              />
           </ThemeProvider>
          </StylesProvider>
        ) : null}
        </React.Fragment>
      )
    } else {
      return <div>Loading...</div>
    }
  }
}

AboutPagePreview.propTypes = {
  entry: PropTypes.shape({
    getIn: PropTypes.func,
  }),
  widgetFor: PropTypes.func,
}

export default AboutPagePreview
```

**Comment by talves** [comment]
@d3sandoval this is close to how I'm handling the previews for `material-ui` also, but you might have a better implementation on the ref. Nice job.

**Comment by homearanya** [comment]
Thanks @d3sandoval, that looks like a nice approach. I like the way the callback ref is used.

 I ended up using the approach that @talves show me on Gitter, which worked for me. Tony, I hope you don't mind that I share it here:

> Basically, I render the full page style context into a string and inject that into a component for the preview. I also do the same for the Preview component.

`PreviewContainer.js`

```javascript
import React from 'react'

export default props => (
  <React.Fragment>
    <style
      type="text/css"
      id={props.id}
      dangerouslySetInnerHTML={{ __html: props.innerStyles }}
    />
    <div
      dangerouslySetInnerHTML={{ __html: props.innerHTML }}
    />
  </React.Fragment>
)
```
`withJss.js`

```javascript
function withJss(Component) {
  class WithRoot extends React.Component {
    constructor(props) {
      super(props);
      this.muiPageContext = getPageContext();
    }

    render() {
      return (
        <JssProvider registry={this.muiPageContext.sheetsRegistry}>
          {/* MuiThemeProvider makes the theme available down the React
              tree thanks to React context. */}
          <MuiThemeProvider
            theme={this.muiPageContext.theme}
            // sheetsManager={this.muiPageContext.sheetsManager}
          >
            {/* CssBaseline kickstart an elegant, consistent, and simple baseline to build upon. */}
            <CssBaseline />
            <Component {...this.props} />
          </MuiThemeProvider>
        </JssProvider>
      );
    }
  }

  return WithRoot;
}

export default withJss;
```
`ServicesPreview .js`

```javascript
const styles = theme => ({
  root: {
    flexGrow: 1,
    marginTop: 5,
  },
  wrapper: {
    padding: 20,
  },
});
const StyledPreview = ({ classes, data }) => (
  <div className={classes.root}>
    <div className={classes.wrapper}>
      <PageHeader data={{title: data.services.title, description: data.services.introduction}} />
    </div>
    <div className={classes.wrapper}>
      <IntroSection data={data} />
      <Services data={data} />
    </div>
  </div>
)

const Preview = withJss(withStyles(styles)(StyledPreview))

export default class ServicesPreview extends React.Component {
  constructor(props) {
    super(props);
    this.muiPageContext = getPageContext()
  }
  render () {
    // const data = this.props.entry.toJS().data // This or next line
    const data = this.props.entry.getIn(['data']).toJS()
    const previewHTML = renderToString(<Preview data={{services: { ...data }}}/>)
    return (
      <PreviewContainer
        innerStyles={this.muiPageContext.sheetsRegistry.toString()}
        innerHTML={previewHTML}
      />
    )
  }
}
```
>`JssProvider` in this case is the `react-jss/lib/JssProvider` component in the jss monorepo.

>that example preview is pretty complicated, but wanted you to see what it is rendering. I have multiple components being rendered into one preview. As you can see, the `muiPageContext.sheetsRegistry` object is created by `import { SheetsRegistry } from 'jss'`

I must say @d3sandoval approach looks simpler

**Comment by talves** [comment]
@homearanya The simpler solution will work fine too, but I have some edge cases that the pre-render is required using the method above. A combination of these both would cover it. 

I would start with the simpler approach and move from there (captain obvious 😜).

I wonder if the cms should be forwarding the ifame ref to the preview, to be able to manage the

**Comment by yoavniran** [comment]
thanks @devolasvegas that solved the issue for me!

created a short gist for my PreviewLayout component using the more appropriate (IMO) useRef & useLayoutEffect based on the example above - 

https://gist.github.com/yoavniran/0953eff3d88f385431b9decc0c3a6be5

showing how to enable styled-components with netlify-cms

**Comment by erezrokah** [comment]
Closing as the issue seems resolved. Please re-open if necessary

**Comment by erquhart** [comment]
This may merit staying open. What we have so far are instructions for individual libraries, most of which are on the tedious side. At a minimum we should document this stuff, at a maximum we should have a simpler support model.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by erezrokah** [comment]
I re-opened an added a `pinned` label so it won't get marked as stale.
I have a suggestion though, we can close this one and open a new issue with the title "Document how to use with CSS in JS libs" and link to this issue for reference.
I think once we start writing the documentation it will make it easier to decide if that is enough or we should make some code changes.

What do you think?

**Comment by bencao** [comment]
Thanks for the previous comments, which inspired me a lot!

## Short Term Solution

In my opinion, the most natural solution is to create a higher-order component that adds support to a specific CSS-IN-JS library.

For example, I would add support for styled-components with the following snippet today:

```
// src/cms/with-styled.js ,  define a higher-order function to support styled

import React from "react";
import { StyleSheetManager } from "styled-components";

export default Component => props => {
  const iframe = document.querySelector("#nc-root iframe");
  const iframeHeadElem = iframe && iframe.contentDocument.head;

  if (!iframeHeadElem) {
    return null;
  }

  return (
    <StyleSheetManager target={iframeHeadElem}>
      <Component {...props} />
    </StyleSheetManager>
  );
};
```

```
// src/cms/cms.js, use the higher-order function defined earlier

import CMS from "netlify-cms-app";
import withStyled from "./with-styled";

import UserPreview from "./preview-templates/UserPreview";
import OrderPreview from "./preview-templates/OrderPreview";

CMS.registerPreviewTemplate("schedule", withStyled(UserPreview));
CMS.registerPreviewTemplate("speakers", withStyled(OrderPreview));
```


## Long Term Solution

But for the long term, in order to achieve best possible user experience, it would be best to hide these dirty details and try to detect whether the project is using `styled-component` or `css-modules` and add support for these CSS-IN-JS libraries automagically, within the lower level `CMS.registerPreviewTemplate` function:

```

function smartRegisterPreviewTemplate(name, component) {
  // check styled-components
  try {
     const styled = require("styled-components");
     
     return registerPreviewTemplate(name, withStyled(component));
  } catch (styledNotFound) {
     // do nothing
  }

  // check css-modules
  try {
     const styled = require("css-modules");
     
     return registerPreviewTemplate(name, withCssModules(component));
  } catch (cssModulesNotFound) {
     // do nothing
  }

  return registerPreviewTemplate(name, component);
}
```

**Comment by emileswain** [comment]
Should this work for nextjs implementations as well? I'm using css modules, and have failed to get any of the examples to work. Nextjs is embeded the styles into the admin page <head> as <style> blocks.

I would expect to see the iframe head element to have shows the styles in a similar way, but i'm not. I imagine its because nextjs works in a different way, but i'm not entirely sure. 

///////////
netlify-cms-app 2.13.3
netlify-cms-app.js:143 netlify-cms-core 2.34.0
netlify-cms-app.js:43 Looking for Netlify CMS Proxy Server at 'http://localhost:8081/api/v1'
netlify-cms-app.js:43 Detected Netlify CMS Proxy Server at 'http://localhost:8081/api/v1' with repo: 'emile.info'
netlify-cms-app.js:43 'editorial_workflow' is not supported by 'local_fs' backend, switching to 'simple'

**Comment by emileswain** [comment]
Ok, So i think this is particular to my issue with NextJS, running in dev with local cms, and very likely i'm just not doing things the right way (i'm not that clued into css.modules and styled components). 

But i simply copied the styles in my header into the iframe head element.  I imagine that i'm using withEmotion wrong in some way.


```
import React from "react";
import { CacheProvider } from "@emotion/core";
import createCache from "@emotion/cache";
import weakMemoize from "@emotion/weak-memoize";

const memoizedCreateCacheWithContainer = weakMemoize(container => {
    // @ts-ignore
    let newCache = createCache({ container });
    return newCache;
});

const withEmotion =  Component => props => {
    const iframe:any = document.querySelector("#nc-root iframe");
    const iframeHeadElem = iframe && iframe.contentDocument.head;

    if (!iframeHeadElem) {
        return null;
    }

    const styles:any = document.querySelectorAll("html>head>style");
    let i;
    for (i = 0; i < styles.length; ++i) {
        const style = styles[i];
        iframeHeadElem.appendChild(style.cloneNode(true));
    }

    return (
        <CacheProvider value={memoizedCreateCacheWithContainer(iframeHeadElem)}>
            <Component {...props} />
        </CacheProvider>
    );
};

export default withEmotion;
```

**Comment by miklschmidt** [comment]
None of these solutions only introduces new problems for me. As soon as i register emotion preview templates, the UI goes to shit. This is enough to break it:
```
CMS.registerPreviewTemplate(
	'pages',
	withEmotion(() => <div></div>),
);
```

using the code from https://github.com/netlify/netlify-cms/issues/793#issuecomment-738031461

![Screenshot from 2021-01-28 16-24-37](https://user-images.githubusercontent.com/751171/106159797-736d0000-6185-11eb-8a36-de26eb80c248.png)

Further more, in editing mode, the cms.css gets inserted into the iframe, instead of the main document, so the editors go to shit as well. If i move that back out, the editors look fine. I can't seem to figure out what to do to make the main UI work. I'm not entirely sure what's happening.

**Comment by wittenbrock** [comment]
I'm in the boat as @miklschmidt - none of the above solutions worked. The CMS's styles look fine until I add this line:

`CMS.registerPreviewTemplate('blog', () => <div>Test</div>);`

Then they have the janky formatting mirrored in his screenshot.

gatsby: 2.32.2
emotion: 11.0.0
gatsby-plugin-emotion: 6.0.0
netlify-cms-app: 2.14.21

**Comment by erezrokah** [comment]
For those who are experiencing the issue in https://github.com/netlify/netlify-cms/issues/793#issuecomment-769165829, can you try downgrading to `emotion@10` and `gatsby-plugin-emotion@4`? The CMS uses `emotion@10`, so the two different version might be conflicting.

**Comment by javialon26** [comment]
I have this working with Emotion 10 but if I try with Emotion 11 the entire preview panel breaks with a React error. Any plan in the near future to update to Emotion 11? Thanks.

**Comment by erezrokah** [comment]
Hi @javialon26, yes there are plans to upgrade but no timeline yet. This will need to be behind a major version bump as it is a breaking change for emotion 10 users. I'm not sure if we should couple this to https://github.com/netlify/netlify-cms/issues/5111 (also a breaking change).

**Comment by mosesoak** [comment]
Hi, we're using the (still beta) variable type lists and yaml aliases to pull off a page-builder model for a project. 

But we're using twin.macro which extends emotion with tailwind functionality. 

Is there any normal way to get these styles working in preview templates yet?

Thanks

**Comment by mosesoak** [comment]
@erezrokah I can put something together if it helps but I'm seeing that Netlify CMS currently doesn't support version 11 of Emotion (the way we use Twin it's basically just emotion with some extra support for tailwind, but I'm not sure we can downgrade to Emotion 10).

Do you know if there's any timeline for getting Netlify CMS updated so it can work with the latest version of Emotion? We are actively using the CMS with client projects now so not having preview is becoming a serious problem for us... thanks!

**Comment by erezrokah** [comment]
Hi @mosesoak, there isn't a timeline for upgrading emotion, however we'd be happy for a contribution.

The upgrade will need to be behind a major version bump as it is a breaking change, see https://github.com/netlify/netlify-cms/issues/5652.

**Comment by mosesoak** [comment]
Thanks @erezrokah -- I ended up taking a different approach since extracting styles from Emotion 11 wasn't working, and also because we are using Gatsby, so even with styling working our components that include special Gatsby functionality like `useStaticQuery` fail to run.

I realized that since we're using a page-builder model where all components for a single page are listed in a variable-type list, that the ideal way to do this would be to show the actual rendered web page, and then overlay the new data changes onto it.

I was able to do it that way! I realize that it doesn't cover all use cases of this CMS but will post the details here in case it's helpful for anyone else.

### "I heard you liked iframes, so I put an iframe in your iframe" 😏

The page is loaded into a second iframe inside the preview, like so:

![image](https://user-images.githubusercontent.com/24345/148294682-46552115-403b-4a0e-a832-10da9d4740f2.png)

At first you're seeing the original page with unaltered content, but as soon as you type into a CMS field you see the content change without any glitches. Since it's precompiled, all the jss styles are included in the page already so no need to use CMS styling. It was cool to find that changing images in the CMS swaps the Gatsby images without any trouble, that's due to a special image component we wrote that will replace the GatsbyImage with a regular img tag if a string src is passed to it.

Data is sent into the secondary iframe using [the secure `postMessage` javascript API](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage) to send a same-origin MessageEvent. The assumption is that this isn't easily spoofable, but our site is static and doesn't have any user session or sensitive data in it so we're happy trusting the browser vendors with this API, and would be happy to hear from anyone here if that's a bad idea.

Because we're doing a page-builder model, it actually works to just pass the entire `entry.toJS().data` blob and use that as a direct replacement for the pageContext blob that Gatsby provides during static generation, which makes this a really lightweight implementation.

```typescript
const PagePreview: FC<PreviewTemplateComponentProps> = ({ entry }) => {
  const win = useRef<HTMLIFrameElement | null>(null);
  
  // home page is a special case without a slug
  let src = `${window.location.origin}/`;
  if (props.slug !== 'home') {
    src += props.slug;
  }
  
  // not in hook - refresh page data on render
  win.current?.contentWindow?.postMessage(props, location.origin);
  
  return (
    <iframe
      ref={win}
      onLoad={handleLoad}
      title={props.slug}
      style={{ width: '100%', height: '100vh' }}
      src={src}
    />
  );
};

CMS.registerPreviewTemplate('pages', PagePreview);
```

Then in the main page template,
```typescript
  // this is subbed in for pageContext when set
  const [cmsPreviewData, setCmsPreviewData] = useState<typeof pageContext>();

  useEffect(() => {
    if (typeof window !== 'undefined') {
      const handleDataChange = (event: MessageEvent) => {
        // Verify same-origin. Browsers won't encode active scripts in postMessage
        // Also lightly validate data structure but keep it minimal to avoid false flags
        if (
          event instanceof MessageEvent &&
          event.origin === window.origin &&
          event.data?.slug === pageContext.slug &&
          Array.isArray(event.data?.['components'])
        ) {
          setCmsPreviewData(event.data);
        } else {
          console.error('invalid message received');
        }
      };
      window.addEventListener('message', handleDataChange);
      return () => window.removeEventListener('message', handleDataChange);
    }
  }, [pageContext]);
```

I added a few other niceties, including a loading message and a `height` state that sets the inner iframe height to the page height on load, which avoids a double scrollbar and lets 'scroll sync' work.

I would hesitate to recommend that anyone use this on a site with live user sessions, but it seems like a great solution for a static site. We haven't made this live yet so please reply with feedback if you have any! Thanks

**Comment by starkys-brzezina** [comment]
Hello,
Have anyone managed to make it work with MUI 5 / emotion 11 🙏 ? Also.. does anyone have at least a slight idea where to start if I wanted to try solving the issue in the core repo?

**Comment by starkys-brzezina** [comment]
> Hi @mosesoak, there isn't a timeline for upgrading emotion, however we'd be happy for a contribution.
> 
> The upgrade will need to be behind a major version bump as it is a breaking change, see #5652.

I'll give it a shot 🙂 When using MUI 5 npm has a lot of issues with dependency conflicts and that's bugging me dearly

**Comment by starkys-brzezina** [comment]
> Hello, Have anyone managed to make it work with MUI 5 / emotion 11 🙏 ? Also.. does anyone have at least a slight idea where to start if I wanted to try solving the issue in the core repo?

Okay so I tried out a couple of solutions proposed here and I can verify that emotion's cache provider with iframe as a container element solves the issue

`import React from 'react';

import { PreviewTemplateComponentProps } from 'netlify-cms-core';

import { CssBaseline, ThemeProvider } from '@mui/material';

import createCache from '@emotion/cache';
import { CacheProvider } from '@emotion/react';

import { BlogPostTemplate } from '../../templates/blog-post';
import theme from '../../theme';

const BlogPostPreview = ({ entry, widgetFor }: PreviewTemplateComponentProps) => {
  const cache = createCache({
    key: 'preview',
    container: document.getElementById('preview-pane').contentWindow.document.head,
  });

  return (
    <CacheProvider value={cache}>
      <ThemeProvider theme={theme}>
        <CssBaseline />
        <BlogPostTemplate
          content={widgetFor('body')}
          description={entry.getIn(['data', 'description'])}
          tags={entry.getIn(['data', 'tags'])?.toJS()}
          title={entry.getIn(['data', 'title'])}
          date={entry.getIn(['data', 'date'])}
          featuredpost={entry.getIn(['data', 'featuredpost'])}
        />
      </ThemeProvider>
    </CacheProvider>
  );
};

export default BlogPostPreview;`

One problem tho, I notices that in preview, the styles are being duplicated (with every mount or render, i'm not sure), any advice appriciated 

<img width="521" alt="Screenshot 2023-01-07 at 0 00 05" src="https://user-images.githubusercontent.com/50992254/211113989-75390ce8-289d-4bbe-a5ed-255ac582fb73.png">

**Comment by starkys-brzezina** [comment]
Turns out memoizing the cache helps

```import React from 'react';

import { PreviewTemplateComponentProps } from 'netlify-cms-core';

import { CssBaseline, ThemeProvider } from '@mui/material';

import createCache from '@emotion/cache';
import { CacheProvider } from '@emotion/react';
import weakMemoize from '@emotion/weak-memoize';

import { BlogPostTemplate } from '../../templates/blog-post';
import theme from '../../theme';

const memoizedCreateCacheWithContainer = weakMemoize((container: any) => {
  return createCache({
    key: 'preview',
    container,
  });
});

const BlogPostPreview = ({ entry, widgetFor }: PreviewTemplateComponentProps) => {
  return (
    <CacheProvider
      value={memoizedCreateCacheWithContainer(
        (document.getElementById('preview-pane') as any).contentWindow.document.head
      )}
    >
      <ThemeProvider theme={theme}>
        <CssBaseline />
        <BlogPostTemplate
          content={widgetFor('body')}
          description={entry.getIn(['data', 'description'])}
          tags={entry.getIn(['data', 'tags'])?.toJS()}
          title={entry.getIn(['data', 'title'])}
          date={entry.getIn(['data', 'date'])}
          featuredpost={entry.getIn(['data', 'featuredpost'])}
        />
      </ThemeProvider>
    </CacheProvider>
  );
};

export default BlogPostPreview;

**Comment by starkys-brzezina** [comment]
> > Hi @mosesoak, there isn't a timeline for upgrading emotion, however we'd be happy for a contribution.
> > The upgrade will need to be behind a major version bump as it is a breaking change, see #5652.
> 
> I'll give it a shot 🙂 When using MUI 5 npm has a lot of issues with dependency conflicts and that's bugging me dearly
> 
> _Edit: WIP, one test suite failing I don't understand why, I'll try to solve it or create the PR anyway_

https://github.com/netlify/netlify-cms/pull/6645 One test suite failing I don't understand why
